src/pyhf/tensor/common.py

```diff
@@ -31,18 +33,23 @@ def stitch(self, data):
         tensorlib, _ = get_backend()
         assert len(self.partition_indices) == len(data)
 
-        data = tensorlib.concatenate(data, axis=-1)
-        data = tensorlib.einsum('...j->j...', data)
-        stitched = tensorlib.gather(data, tensorlib.astensor(self.sorted_indices))
-        stitched = tensorlib.einsum('...j->j...', stitched)
-        return stitched
+        if self.batch_size is None:
+            data = tensorlib.concatenate(data, axis=-1)
+            stitched = tensorlib.gather(data, self.sorted_indices)
+            return stitched
```

Comment:
Instead of multiple `returns` could just return the result of the `ifelse`
```suggestion
        else:
            data = tensorlib.concatenate(data, axis=-1)
            data = tensorlib.einsum('...j->j...', data)
            stitched = tensorlib.gather(data, self.sorted_indices)
            stitched = tensorlib.einsum('...j->j...', stitched)
        return stitched
```